src/common/shared_cache.hpp

```diff
@@ -159,45 +219,54 @@ class SharedLRU {
   void dump_weak_refs(std::ostream& out) {
     for (const auto& [key, ref] : weak_refs) {
       out << __func__ << " " << this << " weak_refs: "
-	  << key << " = " << ref.second
-	  << " with " << ref.first.use_count() << " refs"
+	  << key << " with " << ref.ptr.use_count() << " refs"
 	  << std::endl;
     }
   }
 
   //clear all strong reference from the lru.
   void clear() {
     while (true) {
+#ifndef WITH_SEASTAR
       VPtr val; // release any ref we have after we drop the lock
+#endif
       std::lock_guard locker{lock};
       if (size == 0)
         break;
-
+#ifndef WITH_SEASTAR
       val = lru.back().second;
+#endif
       lru_remove(lru.back().first);
     }
   }
 
   void clear(const K& key) {
+#ifndef WITH_SEASTAR
     VPtr val; // release any ref we have after we drop the lock
     {
       std::lock_guard l{lock};
-      typename map<K, pair<WeakVPtr, V*>, C>::iterator i = weak_refs.find(key);
-      if (i != weak_refs.end()) {
-	val = i->second.first.lock();
+      if (auto i = weak_refs.find(key); i != std::end(weak_refs)) {
+	val = i->second.ptr.lock();
       }
       lru_remove(key);
     }
+#else
+    lru_remove(key);
+#endif
   }
 
   void purge(const K &key) {
+#ifndef WITH_SEASTAR
     VPtr val; // release any ref we have after we drop the lock
+#endif
     {
       std::lock_guard l{lock};
-      typename map<K, pair<WeakVPtr, V*>, C>::iterator i = weak_refs.find(key);
-      if (i != weak_refs.end()) {
-	val = i->second.first.lock();
-        weak_refs.erase(i);
```

Comment:
It looks this was the headache enforcing `Cleanup` to hold key and perform `::find` over `std::map`.